### PR TITLE
Guard boto3_elasticsearch loading properly

### PR DESCRIPTION
**Pull request by oeuftete**
### What does this PR do?

Fixes this traceback on module loading, bringing `boto3_elasticsearch` into line with other `boto3` modules.

```text
2020-09-25 14:27:37,230 [salt.loader      :1764][DEBUG   ][5970] Error loading module.boto3_elasticsearch: __init__ failed
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/salt/loader.py", line 1754, in _load_module
    module_init(self.opts)
  File "/usr/lib/python3.6/site-packages/salt/modules/boto3_elasticsearch.py", line 96, in __init__
    __utils__["boto3.assign_funcs"](__name__, "es")
  File "/usr/lib/python3.6/site-packages/salt/loader.py", line 1278, in __getitem__
    func = super(LazyLoader, self).__getitem__(item)
  File "/usr/lib/python3.6/site-packages/salt/utils/lazy.py", line 108, in __getitem__
    raise KeyError(key)
KeyError: 'boto3.assign_funcs'
```

### What issues does this PR fix or reference?
Fixes (partially): #55848

### Merge requirements satisfied?

- (n/a) Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [ ] Tests written/updated

If tests are required for this, I need guidance.

### CODE REVIEW

**Comment by waynew** [comment]
Should this have a __virtual__, using the `HAS_BOTO`?

**Comment by oeuftete** [comment]
> Should this have a **virtual**, using the `HAS_BOTO`?

:man_shrugging: 

I'm mostly thoughtlessly following the pattern of similar modules, e.g.:

https://github.com/saltstack/salt/blob/ee75b13cc0dce03c9964969c22688710f3346502/salt/modules/boto_cloudwatch_event.py#L77-L87

**Comment by oeuftete** [comment]
It looks like the pre-commit formatting somehow broke the tests.

```
Traceback (most recent call last):
  File "/tmp/kitchen/testing/tests/unit/modules/test_boto3_elasticsearch.py", line 399, in test_add_tags_error
    "testdomain", tags={"foo": "bar", "baz": "qux"}, **CONN_PARAMETERS
  File "/tmp/kitchen/testing/salt/modules/boto3_elasticsearch.py", line 152, in add_tags
    "TagList": [{"Key": k, "Value": value} for k, value in tags or {}.items()],
  File "/tmp/kitchen/testing/salt/modules/boto3_elasticsearch.py", line 152, in <listcomp>
    "TagList": [{"Key": k, "Value": value} for k, value in tags or {}.items()],
ValueError: too many values to unpack (expected 2)
```

File: salt/modules/boto3_elasticsearch.py [hunk 1]
```diff
@@ -156,7 +150,7 @@ def add_tags(
         boto_params = {
             "ARN": arn,
             "TagList": [
-                {"Key": k, "Value": value} for k, value in six.iteritems(tags or {})
+                {"Key": k, "Value": value} for k, value in tags.items() or {}.items()
```
Comment (waynew) [inline salt/modules/boto3_elasticsearch.py:153]:
```suggestion
	              {"Key": k, "Value": value} for k, value in (tags or {}).items()
```
Yeah, that pre-commit horked this line up completely. Something seems off with the indention here? but... this is the logic fix.

Comment (waynew) [reply salt/modules/boto3_elasticsearch.py:153]: cc @s0undt3ch

**Review by waynew — CHANGES_REQUESTED** [review]
That comprehension needs some fixing.

I don't have strong feelings about tests here, but you could easily write two tests that cover both `HAS_BOTO` and `not HAS_BOTO`.

I still think that we want virtual to be returning false without boto though, right? If we don't have boto, we can't connect to elasticsearch... because we don't have boto?

**Comment by waynew** [comment]
oooh, I see what you mean. That check boto reqs probably returns false without boto 👍

File: salt/modules/boto3_elasticsearch.py (hunk 1 above)
Comment (oeuftete) [reply salt/modules/boto3_elasticsearch.py:153]: I think what's in the PR right now is logically ok, but yours is clearer, I'll add that.

What the pre-commit did originally was change `six.iteritems(tags or {})` to `tags or {}.items()`, which is definitely not ok!

Comment (waynew) [reply salt/modules/boto3_elasticsearch.py:153]: what's in the PR now will fail when `tags` is None: 

```
>>> tags = None
>>> tags.items() or {}.items()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'NoneType' object has no attribute 'items'
```

That was an implicit fallback with `iteritems(tags or {})` 🙃 

The other option would be `tags.items() if tags else {}.items()` but that would be silly 😂

Comment (oeuftete) [reply salt/modules/boto3_elasticsearch.py:153]: :facepalm:

Comment (oeuftete) [reply salt/modules/boto3_elasticsearch.py]: I am really confused now.  How did this manage to get in to 3002 in d089e810fe/https://github.com/saltstack/salt/pull/58622?  The reason I saw and fixed this (not quite correctly) in the first place is that it was failing a test.  How did that not happen for the 3002 release?

Comment (oeuftete) [reply salt/modules/boto3_elasticsearch.py]: Ah, I see, the comprehension was changed: https://github.com/saltstack/salt/pull/58622/files#diff-fe8071ebcb4727856197b666dbf829930088a5d5fc5d8c82a0d806d03ecf059bR149

File: salt/modules/boto3_elasticsearch.py [hunk 2]
```diff
@@ -84,7 +85,9 @@ def __virtual__():
 
 def __init__(opts):
     _ = opts
-    __utils__["boto3.assign_funcs"](__name__, "es")
+    if HAS_BOTO:
+        salt.utils.compat.pack_dunder(__name__)
```
Comment (oeuftete) [inline salt/modules/boto3_elasticsearch.py:89]: bad rebase, fixing

**Review by s0undt3ch — APPROVED** [review]

**Comment by ScriptAutomate** [comment]
Is this issue the same one being experienced here:
- https://github.com/saltstack/salt/issues/55187

I had initially added to the conversation in this issue because of all of these traceback error outputs in the debug log for `pkg.installed`, including the one that looks to be getting addressed in this PR

**Comment by oeuftete** [comment]
> Is this issue the same one being experienced here:
> 
> * #55187
> 
> I had initially added to the conversation in this issue because of all of these traceback error outputs in the debug log for `pkg.installed`, including the one that looks to be getting addressed in this PR

@ScriptAutomate I think it's best to keep this separate.  They're sort of related, in that they are both issues where a not-quite-right interaction between `__virtual__` and `__init__` leads to a module load failure with traceback, but the issues are independent.

**Comment by sagetherage** [comment]
@major0 if you want to review or at least know about this PR simply letting you know :)

**Review by Ch3LL — APPROVED** [review]

**Comment by Ch3LL** [comment]
@waynew can you re-review here

**Review by twangboy — APPROVED** [review]

**Review by waynew — APPROVED** [review]

**Pull request merged**